Allow --hostname with --net=host

jheiss opened the pull request:
Docker creates a UTS namespace by default, even with --net=host, so it is
reasonable to let the user set the hostname. Note that --hostname is
forbidden if the user specifies --uts=host.

Closes #12076
Signed-off-by: Jason Heiss jheiss@aput.net

jheiss commented on docs/reference/run.md:
```diff
@@ -237,15 +237,13 @@ $ docker run -it --rm --pid=host myhtop
 hostname of the container to change as the hostname of the host changes.  A
 more advanced use case would be changing the host's hostname from a container.
 
-> **Note**: `--uts="host"` gives the container full access to change the
```
I removed this because even with --uts=host you can't set the hostname as you lack the SYS_ADMIN capability

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/832/console):

~~~console

---
ed1d3d.
+ ec=125
+ set +x

----------------------------------
ERROR: Failed to build test binary
----------------------------------



-----------------------------------------------
ERROR: Failed with exitcode 125 at Wed Feb 10 14:47:00 CUT 2016.
-----------------------------------------------


INFO: Tidying up at end of run
INFO: Nuking /d/CI
INFO: Zapped successfully
INFO: End of cleanup
INFO: Ended at Wed Feb 10 14:47:01 CUT 2016 (1m 15s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe D:\temp\hudson8181777298685205788.sh
+ set +e
+ set +x
INFO: End of cleanup
Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
---

~~~

estesp commented:
Given UTS is separate from network namespace, I think this is agreeable.  A user has to understand that this allows them to have some inconsistency between `/etc/hosts` + `/etc/hostname` and the "new" UTS hostname..so not sure if we need more documentation there..

jfrazelle commented:
what is the usecase for having this, I think it is like this to prevent confusion, majority of people believe uts and net are the same thing and if you are going to ping say my hostname debian from a net host container you would expect it to work

jheiss commented:
My particular use case is running the MIT Kerberos replication server (kpropd) in a container. The MIT Kerberos replication protocol uses a special "secure" Kerberos mode that does some extra sanity checking involving the hostname and IP address. In order to comply with these checks I have to run the container with --net=host but set the hostname to a value that differs from the host. To do that currently I have to give the SYS_ADMIN capability to the container so that I can run the hostname command as part of my entrypoint. If I could use --hostname I wouldn't have to give the container SYS_ADMIN.

thaJeztah commented:
ping @estesp @jfrazelle what's the status here, what are your current thoughts?

jfrazelle commented:
Ok I think this is all good, just can we keep the current default behavior, _but_ if they pass --hostname then it will set it over that, also a note in the docs maybe explaining how and why this works without changing the hostname on their host would be good just so people understand a bit better

estesp commented:
@jheiss I think we agreed this is reasonable and @jfrazelle asked for a bit more clarification in the docs (I think under the `--net=host` section; maybe instead of only remove `--hostname` as a conflicting param add a comment about the fact that you can set `--hostname` because the UTS namespace is unique in the container and does not affect the hostname on the host.

I've moved this to code review as I think we agreed this is a reasonable change at last week's review session.

jheiss commented:
git and I clearly don't get along. Sigh.

jheiss commented:
Give me a bit to see if I can undo my failed attempt to squash/rebase

moxiegirl commented:
@jheiss At this point, you might be better off doing this:

1. Copy the file(s) with your actual change outside of the repo.
2. Close this PR.
3. Back in your local clone, fetch
4. Rebase to docker/docker master branch
5. Create a new branch
6. Copy your changes into your new branch
7. Create a new PR

jheiss committed:
Allow --hostname with --net=host

Docker creates a UTS namespace by default, even with --net=host, so it
is reasonable to let the user set the hostname. Note that --hostname is
forbidden if the user specifies --uts=host.

Closes #12076
Signed-off-by: Jason Heiss <jheiss@aput.net>

jheiss commented:
OK, I think my attempt at git surgery was successful.

I added a sentence to the host network section to try to better explain the interaction. Let me know if anyone has suggestions. For example, should that be a separate paragraph? Called out as a bolded note? (It doesn't seem to me to warrant that much attention.) Does it need to be more verbose?

In re-reading the documentation I noticed that with my change in the UTS section we now the sentence I added that says --hostname is forbidden with host UTS mode, and then a few sentences later it says that an advanced use case would be to set the actual hostname from within the container. That seems potentially confusing without further explanation of how you would go about that (giving the container SYS_ADMIN capability and running `hostname` in the container). Not sure if I should try to explain that better, or remove the note about setting the actual hostname from within the container? Or just leave it alone and assume advanced users will figure it out?

thaJeztah commented:
WindowsTP4 timed out after 2 hours :cry: https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/2396/console

I'll restart that job

```
03:47:44 SKIP: docker_cli_search_test.go:11: DockerSuite.TestSearchOnCentralRegistry (Test requires a Linux daemon)
03:47:45 panic: test timed out after 2h0m0s
```

moxiegirl commented:
@jheiss Thank you for sorting that. LGTM.

thaJeztah commented:
docs LGTM

thaJeztah commented:
ping @estesp @jfrazelle does code LGTY, if it does, we can merge this :heart:

vdemeester commented:
LGTM 🐰

estesp commented:
looks good, thanks @jheiss! LGTM
